frontend/components/admin/Photos.jsx

```diff
@@ -48,50 +50,62 @@ class Photos extends React.Component {
+                const reader = new FileReader();
+                reader.onload = async () => {
+                    const body = {
+                        imageName: this.state.album[index].name,
+                        imageData: reader.result,
+                        albumAddress: this.state.albumName,
+                    };
+                    const res = await uploadAlbumPhoto(body);
+                    if (res === false) {
+                        await this.setState({
+                            skipped: this.state.skipped + "Failed " + image.name + " upload\n",
+                        });
+                    }
+                    resolve();
+                };
+                reader.readAsDataURL(image);
+            });
+
+            await filePromise;
+        }
+        await this.setState({ progress: this.state.progress + 1 });
+    };
+
     submitButton = async () => {
         await this.setState({ uploading: true });
-        console.log(this.state.selectedPhotoAlbum);
         if (this.state.selectedPhotoAlbum === "") {
             await createPhotoAlbum({ name: this.state.albumName, date: new Date() });
         }
+        let currentPromises = [];
         for (let i = 0; i < this.state.album.length; i++) {
-            await this.setState({ progress: i });
-            const image = this.state.album[i];
-            if (
-                image.type !== "image/jpeg" &&
-                image.type !== "image/png" &&
-                image.type !== "image/gif" &&
-                image.type !== "image/tiff" &&
-                image.type !== "image/webp"
-            ) {
-                await this.setState({
-                    skipped: this.state.skipped + "Skipped " + image.name + "\n",
-                });
+            if (currentPromises.length < NO_PARALLEL_UPLOADS) {
+                currentPromises.push(this.uploadSingleImage(i));
             } else {
-                const filePromise = new Promise((resolve) => {
-                    const reader = new FileReader();
-                    reader.onload = async () => {
-                        console.log(this.state.album[i].name);
-                        const body = {
-                            imageName: this.state.album[i].name,
-                            imageData: reader.result,
-                            albumAddress: this.state.albumName,
-                        };
-                        const res = await uploadAlbumPhoto(body);
-                        if (res === false) {
-                            await this.setState({
-                                skipped: this.state.skipped + "Failed " + image.name + " upload\n",
-                            });
-                        }
-                        resolve();
-                    };
-                    reader.readAsDataURL(image);
-                });
-
-                await filePromise;
+                const [promiseToRemove] = await Promise.race(
```

Comment:
What does lines 100-104 do?

Reply:
So basically:
- `Promise.race` returns the result of the first promise to return (in this case the promises are waiting for an image to upload)
- the promises are stored in a list of promises. we map the promises to return the the promise once completed, so we have the promise we want to remove after the .race is complete.
- We want to remove the first promise to complete from the list, so we splice `currentPRomises` to remove `promiseToRemove`. 
- Now we have fewer than 5 promises in `currentPromises`, so we start uploading the next image and push that to the list of promises.

In a single sentence, these lines wait for one of the currently uploading images to finish uploading, then starts uploading another image. This means that there are constantly 5 images uploading at a single time, and never more.

I've added a comment above to make the what these lines are doing slightly clear 😊